cmd/notation/internal/plugin/plugin.go

```diff
@@ -0,0 +1,179 @@
+	mgr := plugin.NewCLIManager(dir.PluginFS())
+	plugin, err := mgr.Get(ctx, pluginName)
+	if err != nil {
+		return nil, err
+	}
+	return plugin.GetMetadata(ctx, &proto.GetMetadataRequest{})
+}
+
+// GetPluginMetadata returns plugin's metadata given plugin path
+func GetPluginMetadata(ctx context.Context, pluginName, path string) (*proto.GetMetadataResponse, error) {
+	plugin, err := plugin.NewCLIPlugin(ctx, pluginName, path)
+	if err != nil {
+		return nil, err
+	}
+	return plugin.GetMetadata(ctx, &proto.GetMetadataRequest{})
+}
+
+// ComparePluginVersion validates and compares two plugin semantic versions
+func ComparePluginVersion(v, w string) (int, error) {
+	// semantic version strings must begin with a leading "v"
+	if !strings.HasPrefix(v, "v") {
+		v = "v" + v
+	}
+	if !semver.IsValid(v) {
+		return 0, fmt.Errorf("%s is not a valid semantic version", v)
+	}
+	if !strings.HasPrefix(w, "v") {
+		w = "v" + w
+	}
+	if !semver.IsValid(w) {
+		return 0, fmt.Errorf("%s is not a valid semantic version", w)
+	}
+	return semver.Compare(v, w), nil
+}
+
+// ValidateCheckSum returns nil if SHA256 of file at path equals to checkSum.
+func ValidateCheckSum(path string, checkSum string) error {
+	rc, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+	dgst, err := digest.FromReader(rc)
+	if err != nil {
+		return err
+	}
+	enc := dgst.Encoded()
+	if enc != strings.ToLower(checkSum) {
+		return fmt.Errorf("plugin checksum does not match user input. Expecting %s", checkSum)
+	}
+	return nil
+}
+
+// DownloadPluginFromURL downloads plugin file from url to a tmp directory
+func DownloadPluginFromURL(ctx context.Context, url string, tmpFile *os.File) error {
+	// Get the data
+	client := getClient(ctx)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	// Check server response
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("bad status: %s", resp.Status)
+	}
+	// Write the body to file
+	lr := io.LimitReader(resp.Body, maxPluginSourceBytes)
+	_, err = io.Copy(tmpFile, lr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getClient returns an *auth.Client
+func getClient(ctx context.Context) *auth.Client {
+	client := &auth.Client{
+		Cache:    auth.NewCache(),
+		ClientID: "notation",
+	}
+	client.SetUserAgent("notation/" + version.GetVersion())
+	setHttpDebugLog(ctx, client)
+	return client
+}
+
+// setHttpDebugLog sets http debug logger
+func setHttpDebugLog(ctx context.Context, authClient *auth.Client) {
+	if logrusLog, ok := log.GetLogger(ctx).(*logrus.Logger); ok && logrusLog.Level != logrus.DebugLevel {
+		return
+	}
+	if authClient.Client == nil {
+		authClient.Client = http.DefaultClient
+	}
+	if authClient.Client.Transport == nil {
+		authClient.Client.Transport = http.DefaultTransport
+	}
+	authClient.Client.Transport = trace.NewTransport(authClient.Client.Transport)
+}
```

Comment:
Is it possible to refactor this code as it is duplicated with

https://github.com/notaryproject/notation/blob/b75827cc02aecbde8ec7a6b6b13f10c223ff47e6/cmd/notation/registry.go#L125

Reply:
Yeah, `setHttpDebugLog` should really belong to the `github.com/notaryproject/notation/internal/trace` package. Moved it there and reusing it in `main` and `plugin` packages.